### ISSUE TITLE
FB Resolution Performance Overheads.

### ISSUE THREAD

**Issue by AmbientMalice** [gonetz/GLideN64#583]
Across the board, enabling high resolution modes causes GLideN64's performance to drop dramatically. Excuse my noobness, but is this normal? I know copying FB to RDRAM has some overheads, but these overheads are so bad they effectively make GLideN64 useless if the FB resolution goes above 320x240 unless your CPU is bleeding edge.

For example, Indiana Jones runs at full speed (mostly) with Expansion Pack disabled. (280x200, IIRC.) Enabling it (400x440) causes a 20-30% drop in performance. Maybe worse. Vigilante 8 is painfully slow with FB enabled, and enabling higher resolution modes makes it even worse.

I'm using an Nvidia GTX 750Ti, for reference.

**Comment by LegendOfDragoon** [comment]
>Excuse my noobness, but is this normal?

Yes. That's just the way it is.

**Comment by Nes-player4Life** [comment]
@LegendOfDragoon 
Please make your post(s) useful in some way.

@AmbientMalice 
It almost sounds like a RAM issue. How much memory do you have?

**Comment by AmbientMalice** [comment]
@Nes-player4Life 8GB RAM. 2GB VRAM. CPU is a Core Quad @ 3Ghz. I'm aware that copying to RDRAM is resource intensive. (It was always a weak point for Dolphin, for example.) But Gonetz recently made a commit to improve RDRAM copy performance. https://github.com/gonetz/GLideN64/commit/68941f6cbe78e9384b499bb18b17297c069d95c4 (And I think this may be reflected in World Driver Championship's recent speed boost.)

End on the day, enabling FB in GLideN64 still carries an insane performance hit that scales with FB resolution. (Some games seem worse affected than others.) Now perhaps this is merely a side effect of GLideN64's overall slowish performance. Maybe optimisations to the plugin will make the hefty FB overheads tolerable.

There is *something* wrong with GLideN64's FB code in other areas. Shaking screen when copying FB to texture, for example. And broken depth buffer stuff. But this atrociously slow FB performance risks torpedoing the plugin. Resident Evil 2 won't run with GLideN64 yet, but I dread to think how it will handle the crazy random resolutions of the game's high resolution mode, since it runs flailing into a wall, before hobbling along at 25Hz, when Factor 5 games try to run at 400x440 or 500x200-ish or some such.

Unless of course there is some GPU-specific weirdness going on.

**Comment by theboy181** [comment]
Buy new HW. JUST KIDDING!!

Are you talking about all games or a few?

**Comment by AmbientMalice** [comment]
>Are you talking about all games or a few?

Perfect Dark in High Res mode runs at 40VIs with infra red goggles in places instead of 60VIs minimum. Vigilante 8 runs terribly in normal resolution with FB on, and gets worse with increased resolutions. Episode I - Racer has big performance issues with FB enabled. Pretty much any game with a high resolution mode takes a massive hit if it is enabled. Rogue Squadron/Naboo/Infernal Machine are noticably slower with the expansion pack enabled.

Bear in mind this is completely different to performance issues from raising the plugin's resolution -- I fixed those by forcing Performance mode in Nvidia's control panel.

I guess the underlying problem is N64 games lean heavily on Framebuffer to RAM effects. It's not like Dolphin where, it seems, Framebuffer to Texture works just fine in most cases.

I have no clue what possible steps could be taken to make FB to RAM faster, but in its current state, it's gonna hobble GLideN64 on anything less than bleeding edge hardware.

**Comment by purplemarshmallow** [comment]
But copy fb to RDRAM does not work at all in LLE #554 
I think copying to RDRAM is not the biggest problem here, the slowdowns are caused by something else

**Comment by AmbientMalice** [comment]
@purplemarshmallow Excellent point. I'll do some benchmarking tomorrow to establish whether changing FB settings has any performance effects when running in LLE. I was aware that depth buffer copies were failing in LLE, but wasn't sure about FB copies. Also, I'm not sure whether FB copies are having any performance effect regardless of them failing.

What I do know is that changing FB resolution in LLE affects performance. Maybe RDRAM copy overheads aren't the real problem.

**Comment by LegendOfDragoon** [comment]
If someone could help me with compiling this plugin, I'd be able to benchmark.

**Comment by AmbientMalice** [comment]
>If someone could help me with compiling this plugin, I'd be able to benchmark.

I haven't been able to compile it myself. Too many dependencies and obtuse errors. (EmuCR managed to get working builds. Shame they don't make guides.)

Gonetz was asked about building on Windows. He said

>freetype, boost, qt, pnglib, zlib, qt plugin for Visual Studio.
if you need build, which works not only on your PC, you need static build of all libs, including static qt build from sources. I don't have time to write proper build instruction atm, sorry.

**Comment by purplemarshmallow** [comment]
First you need to open GLideNHQ.sln and compile debug or release. It should build without problems, boost is removed and the other dependencies are included. Then you open GLideN64.sln (I think there should be only one .sln file for everything?). To build the mupen64plus version you need to add freetype to the inlude and library directories. I haven't compiled it myself, I just took it from mupen64plus https://github.com/mupen64plus/mupen64plus-win32-deps. The mupen64plus build works for me.

To build the zilmar spec version you also need to build GLideNUI.lib and you need Qt libraries to build it. I haven't looked into that. Gonetz also uploaded it here https://github.com/gonetz/GLideN64/issues/458. I tried it and it builds but the plugin just crashes.

**Comment by AmbientMalice** [comment]
I am puzzled. Deeply puzzled. It appears that the problem is not always RDRAM related.

LLE is utterly batty. Indiana Jones is highly inconsistent about the intro framerate, especially between 1.1 and current builds. But what is blindingly clear is that enabling expansion pack in current builds drops the intro framerate by about 10VIs from 40ish to 30ish even with the framebuffer completely disabled. So maybe enabling\disabling FB doesn't work properly in LLE?

Perfect Dark's wacky goggle slowdown? Not RDRAM related at all. Note the following images, all taken with FB to RDRAM disabled. (Yes, I made sure to turn off automated settings.)

Goggles off (High Resolution).
![no_fb_off](https://cloud.githubusercontent.com/assets/10704701/8274467/06c2159c-18d6-11e5-8047-9fe6bc878179.jpg)

Goggles on (High Resolution)
![no_fb_on](https://cloud.githubusercontent.com/assets/10704701/8274483/26f9dd72-18d6-11e5-9d54-ffa9a58dcaa4.jpg)

Goggles on (low resolution)
![no_fb_on_low_res](https://cloud.githubusercontent.com/assets/10704701/8274486/32c54aec-18d6-11e5-9b6d-28f355193640.jpg)

Vigilante 8, however, is clearly hit hard by RDRAM.

RDRAM copies off.
![vig8_no_rdram](https://cloud.githubusercontent.com/assets/10704701/8274489/3ddc3d14-18d6-11e5-85b2-ffe2335e1cde.jpg)

RDRAM on (Low Resolution)
![vig8_rdram_low](https://cloud.githubusercontent.com/assets/10704701/8274492/4552b474-18d6-11e5-9534-95bb5ac2e6c0.jpg)

RDRAM on (Medium Resolution)
![vig8_rdram_medium](https://cloud.githubusercontent.com/assets/10704701/8274500/55a54cba-18d6-11e5-9ab1-a00804aa8054.jpg)

In a nutshell, Perfect Dark nosedives into framerate oblivion despite not doing any FB to RDRAM copies.

**Comment by LegendOfDragoon** [comment]
Thanks for the advice @purplemarshmallow  & @AmbientMalice . I'll give it another try. Pretty burnt out, trying to compile and test different plugins :disappointed: . I also want to compile z64gl with MSVC so I can debug that as well.

Interestingly in SD Hiryuu no Ken Densetsu, there's this menu that's mega slow in HLE, yet it's faster in LLE. I thought that was pretty weird ;/ .

**Comment by gonetz** [comment]
Regarding performance hits with copy buffers to RDRAM. Data copied from video memory to PC RAM. Pixel pipeline is stalling during the copy. It's bad for pipeline being in stall. Amount of data depends of game's internal resolution. Copy of 320x240 buffer will be 4 times faster than 640x480 one. Thus, it is not recommended to use "copy buffers" options for games, which use hi-res.

There is an optimization, which works for some games: use two pixel buffers and assync buffer read. If pause between buffer reads is large enough, video buffer can be read without overheads. In our case it is pause between frames. However, in this case we will have copy of previous frame. It is not very noticeable for TV effects or pause screen, but for some games we will get just black or wrong screen instead. I removed that optimization because it does not always work, but I can return it back and make it optional.

**Comment by AmbientMalice** [comment]
>There is an optimization, which works for some games: use two pixel buffers and assync buffer read. If pause between buffer reads is large enough, video buffer can be read without overheads. In our case it is pause between frames. However, in this case we will have copy of previous frame. It is not very noticeable for TV effects or pause screen, but for some games we will get just black or wrong screen instead. I removed that optimization because it does not always work, but I can return it back and make it optional.

I suppose you could add it as an ini setting called RDRAM Speed Hack are something, and just make sure it is thoroughly tested for each game before being enabled by default.

Anyway, @gonetz do you have any idea why Perfect Dark's goggle effects are so resource-intensive despite not using RDRAM copies?

**Comment by gonetz** [comment]
I have performance issues with IR scanner only if in-game hi-res mode is on. Otherwise the game runs smooth fullspeed with IR googles on in full-hd with 8x MSAA. My video card is far from top.

I don't know why hi-res mode causes performance hit. Internal game resolution is not important for plugin's work.

**Comment by purplemarshmallow** [comment]
Here's a suggested fix for Perfect Dark: #704
